classes/base_model.py

```diff
@@ -1,30 +1,65 @@
 import torch
 import classes.utils as utils
+import torch.nn as nn
+import torch.optim as optim
+from classes.interfaces import BaseModel
 
+'''
+This is "class factory" function that generates a new class for each model
+It uses typeclass construction primarily for convinience, but essentially
+works the same way as normal class construction
 
-class base_model():  # TODO class names should use CamelCase convention
-    """
-    params:
-        feed_forward: The feed forward method of a model
-        training_method: The training loop of a model
-        evaluate: Method for running a models feed_forward and parsing the output to UCI notation
-        use_cuda: a boolean whether to use CUDA, only works if CUDA is installed
-        resume: flag to say if we are resuming training
-        filepath: file where to save the model
-    """
+Params
+ff: feed_forward. Must be a class for autograd/cuda to work
+tr: The training function. This is a function, not a class.
+    Params are instantiated in here
+pa: Currently the supervised evaluate function, we'll need
+    to change some things for perform_action
+'''
 
-    def __init__(self, feed_forward, training_method, evaluate, use_cuda=True, resume=False, filepath='auto'):
+
+def generate_class(ff, tr, pa):
+    '''
+    Setting BaseModel doesn't actually do anything atm because we
+    explicitely define the abstract methods in here
+    I don't think it really matters though because explicitely defining the
+    methods should prevent any inheritance issues
+    '''
+    superclasses = (BaseModel,)
+
+    '''
+    I felt that defining training params inside the model was cleaner and
```

Comment:
Just for clarification, which parameters are we talking about?  Is it the parameters of the model (e.g. number of layers, types of layers, etc) or is it metaparameters (e.g. learning rate, optimizer etc)?  If it's the latter then I suggest we do not keep parameters with the model as those are logically separate things.

Reply:
We talked about this a lot in the meeting today. Basically I think calling it a model class is a misnomer and should instead be thought of as a container for a single experiment. Metaparameters are logically separate from the model, but in the context of each generated class being a container for a single experiment I think it makes more sense to store metaparameters inside that container. Sorry for the naming confusion.